### PR TITLE
Overlord jnyman

### PR DESCRIPTION
**Pull request by jnyman**
To run the server and see the application itself:

`rake serve`

To run the Cucumber tests, the RSpec tests, and Rubocop, just run:

`rake`

This was done BDD/TDD first. Coming in as a test engineer, the exercise was a bit too defocusing in terms of the interplay of code and testing. There's a LOT of considerations that start to come up when you bring in Cucumber (which is arguably a very burdensome abstraction layer), not least of which is determining when you are duplicating coverage as well as the appropriate way to structure TDL/BDL (Test Description Language / Business Description Language).

I used Capybara (which is not as flexible as Watir by a long shot). I also used the page object pattern, using the SitePrism gem. I did only the barest implementation of this and provided my own context factory since Capybara, perhaps strangely, does not provide one.

There is a `NOTES.md` file that shows some of my thinking as I went through the exercise.

### CODE REVIEW

File: overlord/features/bomb_usage.feature [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+    And the bomb display indicates the activation code was invalid
+
+    Examples:
+      | code   |
+      | ""     |
+      | "aaaa" |
+      | "1ab3" |
+      | "12"   |
+
+  @javascript
```
Comment (magicalbanana) [inline overlord/features/bomb_usage.feature]: WATIR! :dancer: :bee:

File: overlord/features/bomb_usage.feature [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+    When an incorrect activation code is entered
+    Then the bomb will display as inactive
+    And the bomb display indicates the activation code was incorrect
+
+  @javascript
+  Scenario: Bombs cannot be deactivated with incorrect deactivation codes
+    Given an activated bomb
+    When an incorrect deactivation code is entered
+    Then the bomb will display as active
+    And the bomb display indicates the deactivation code was incorrect
```
Comment (magicalbanana) [inline overlord/features/bomb_usage.feature]: What are your thoughts on having indents on And steps?

File: overlord/features/bomb_usage.feature (hunk 1 above)
Comment (jnyman) [inline overlord/features/bomb_usage.feature]: I know! :) This was before I realized we could branch out a bit. My CYOA will be WATiR based.

File: overlord/features/support/context_factory.rb [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    def on_visit(definition, &block)
+      on(definition, true, &block)
+    end
+
+    # Creates a definition context for actions.
+    #
+    # @param definition [Class] the name of the page class
+    # @param visit [Boolean] true if the definition needs to be called into view
+    # @param block [Proc] logic to execute in the context of the definition
+    def on(definition, visit = false, &block)
```
Comment (magicalbanana) [inline overlord/features/support/context_factory.rb]: I can see PageObject pattern emerging here. :dancer: :bee:

File: overlord/features/support/pages/bomb_page.rb [hunk 4]
```diff
@@ -0,0 +1,44 @@
+class BombPage < SitePrism::Page
+  element :trigger_code, 'input[id=trigger-code]'
+  element :bomb_state, 'input[id=change-bomb-state]'
+  element :notice, 'label[id=notice]'
+  element :bomb_status, 'caption[id=bomb-status]'
+  element :timer_display_value, 'input[id=timer-output]'
+  element :timer_value, 'input[id=timer-countdown]', visible: false
+  element :timer_active, 'input[id=timer-started]', visible: false
+  element :detonation, 'img[id=detonation]'
```
Comment (magicalbanana) [inline overlord/features/support/pages/bomb_page.rb]: Wooo. PageObject! :dancer: :bee:

**Comment by magicalbanana** [comment]
Ich bin beeindruckt! :dancer: :bee:

All is good with this, one thing is I want to know were you stand between DAMP or DRY testing, I am talking about the steps.

I tend to lean toward DAMP first then DRYing it up later on.

**Comment by jnyman** [comment]
I do believe in intent-revealing, expressive tests, thus I put emphasis on DAMP first. (I believe this is what makes testing lucid, which I wrote about here: http://testerstories.com/2015/05/lucid-testing-and-description-languages/.) The DRY aspects, as you say, can come later.

**Comment by magicalbanana** [comment]
Me too, I like to write the DAMP steps first then if I can see a pattern for reusage, that's when I DRY up.

Either way, your PR is good. This can be closed.

File: overlord/NOTES.md [hunk 5]
```diff
@@ -0,0 +1,137 @@
+## Design Notes and Thoughts
```
Comment (paulghaddad) [inline overlord/NOTES.md]: Awesome documentation!

File: overlord/features/bomb_provision.feature [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+    Examples:
+      | code    |
+      | "12345" |
+      | "aaaa"  |
+      | "1ab3"  |
+      | "12"    |
+
+  @javascript
+  Scenario Outline: Custom countdown values must be a whole number of seconds
+    Given the bomb provisioning page
```
Comment (paulghaddad) [inline overlord/features/bomb_provision.feature]: These are fantastic high-level step definitions. This is exactly what we're looking for! Nice work!

**Comment by paulghaddad** [comment]
This is a great Overlord submission. Excellent feature tests; they are right on target.

**Pull request closed without merging**